### PR TITLE
Split RolloverStep into Wait and Action steps

### PR DESCRIPTION
**Pull request by AthenaEryma**
RolloverAction will now periodically check the rollover conditions using
the Rollover API with the dry_run option as an AsyncWaitStep, then run
the rollover itself by calling the Rollover API with no conditions,
which will always roll over, as an AsyncActionStep. This will resolve
race condition issues in policies using RolloverAction.

Replaces https://github.com/elastic/elasticsearch/pull/35391 as this approach is more streamlined.

Relates to https://github.com/elastic/elasticsearch/issues/35313

Also relates to https://github.com/elastic/elasticsearch/issues/35244

---

Background for this PR: We agreed that it would be best to be able to separate checking the rollover conditions from executing the rollover.  In the middle of refactoring the API to accomplish this, I noticed that this is already possible: Calling the Rollover API without providing conditions will unconditionally roll over the index, without using any hacky workarounds such as `max_size: 0b`.  This is not mentioned in the Rollover API documentation - I will be opening a separate PR to add it to the docs.

### CODE REVIEW

File: x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/indexlifecycle/TimeSeriesLifecycleActionsIT.java [hunk 1]
```diff
@@ -231,39 +230,6 @@ public void testRolloverAction() throws Exception {
         assertBusy(() -> assertTrue(indexExists(originalIndex)));
     }
 
-    public void testRolloverAlreadyExists() throws Exception {
```
Comment (gwbrown) [inline x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/indexlifecycle/TimeSeriesLifecycleActionsIT.java]: Whoops - I need to reinstate this integration test - I'll do this soon.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/RolloverAction.java [hunk 2]
```diff
@@ -132,18 +132,23 @@ public boolean isSafeAction() {
 
     @Override
     public List<Step> toSteps(Client client, String phase, Step.StepKey nextStepKey) {
+        StepKey rolloverReadyStepKey = new StepKey(phase, NAME, WaitForRolloverReadyStep.NAME);
+        StepKey rolloverStepKey = new StepKey(phase, NAME, RolloverStep.NAME);
         StepKey updateDateStepKey = new StepKey(phase, NAME, UpdateRolloverLifecycleDateStep.NAME);
-        RolloverStep rolloverStep = new RolloverStep(new StepKey(phase, NAME, RolloverStep.NAME), updateDateStepKey, client,
+
+        WaitForRolloverReadyStep waitForRolloverReady = new WaitForRolloverReadyStep(rolloverReadyStepKey, rolloverStepKey, client,
```
Comment (colings86) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/RolloverAction.java]: Nit: Can we name the key variable and the step variable similarly?

**Review by colings86 — DISMISSED** [review]
Can we re-enable the `ChangePolicyforIndexIT#testChangePolicyForIndex` test as part of this PR so we ca be sure that this fixes the original issue?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/RolloverAction.java (hunk 2 above)
Comment (gwbrown) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/RolloverAction.java]: Done.

**Review by colings86 — COMMENTED** [review]
This LGTM but I'd like @dakrone to take a look too as a second pair of eyes before we merge

**Review by dakrone — APPROVED** [review]
LGTM too, I left one minor comment

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/WaitForRolloverReadyStep.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+ */
+public class WaitForRolloverReadyStep extends AsyncWaitStep {
+
+    public static final String NAME = "check_rollover_ready";
+
+    private static final Logger logger = LogManager.getLogger(WaitForRolloverReadyStep.class);
+
+    private ByteSizeValue maxSize;
+    private TimeValue maxAge;
+    private Long maxDocs;
```
[the remark this answers is not in the document]
Comment (gwbrown) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/WaitForRolloverReadyStep.java]: Can and did!

**Pull request merged**